packages/react-instantsearch/src/core/createConnector.js

```diff
@@ -63,24 +63,41 @@ export default function createConnector(connectorDesc) {
 
         const { ais: { store, widgetsManager } } = context;
         const canRender = false;
+
+        const initialUiState = connectorDesc.getInitialUiState
+          ? connectorDesc.getInitialUiState(props)
+          : {};
+
         this.state = {
-          props: this.getProvidedProps({ ...props, canRender }),
+          uiState: initialUiState,
+          props: this.getProvidedProps({
+            uiState: initialUiState,
+            props: {
+              ...props,
+              canRender,
+            },
+          }),
           canRender, // use to know if a component is rendered (browser), or not (server).
         };
 
         this.unsubscribe = store.subscribe(() => {
           if (this.state.canRender) {
-            this.setState({
+            this.setState(prevState => ({
```

Comment:
is this.state strictly equivalent to prevState here? Just wondering if this can break in any way

Reply:
this.state is actually the wrong way of doing it, since `setState` calls can be batched, and so it could be possible that `this.state` is the value from a previous render (especially when async rendering will be enabled)

Reply:
No it won't be strictly equivalent in some cases. But it should not be a problem, because the `canRender` value change only once (when the component `didMount`). When the condition is `true` we are sure that the `prevState` value is correct since it's the one that is the most up to date.